### ISSUE TITLE
Dragdrop tool throws an exception when trying to reorder ListBox items that are themselves collections.

### ISSUE THREAD

**Issue by BobSammers** [punker76/gong-wpf-dragdrop#309]
## What steps will reproduce this issue?
Using this tool to reorder a `Listbox` whose items are collection objects (for example, deriving from `List<>`) causes an exception.:
- System.ArgumentException: 'The value "TestDDFault.SubMember" is not of type "TestDDFault.ListMember" and cannot be used in this generic collection.'

There is no stack trace.

The dragged item must have members in its collection.

### Expected outcome

Objects are rearranged: no exception

### Repro
This is a complete example, excluding only the code generated by VS when it creates a new project and the project and XAML references to `GongSolutions.Wpf.DragDrop`.

When using this example, `Item 2` must be dragged past `Item 1`, as `2` is the one with sub-members.
```cs
public class ViewModel
{
    public ObservableCollection<ListMember> Members { get; set; } = new ObservableCollection<ListMember>();

    public ViewModel()
    {
        var listMember = new ListMember() { Name = "Item 1" };
        Members.Add(listMember);

        listMember = new ListMember() { Name = "Item 2" };
        listMember.Add(new SubMember() { Name = "Subitem1" });

        Members.Add(listMember);
    }
}

public class ListMember : List<SubMember>
{
    public string Name { get; set; }
}

public class SubMember
{
    public string Name { get; set; }
}
```
```xml
<Window.DataContext>
    <local:ViewModel />
</Window.DataContext>
<Grid>
    <ListBox
        ItemsSource="{Binding Members}"
        DisplayMemberPath="Name"
        dd:DragDrop.IsDragSource="True"
        dd:DragDrop.IsDropTarget="True"
        />
</Grid>
```
### Environment

- GongSolutions.WPF.DragDrop __v1.1.0__ & __v2.0.0-alpha0106__
- Windows OS __10 Pro x64 1803__
- Visual Studio __2017 Community__
- .NET Framework __4.7__

**Comment by punker76** [comment]
@BobSammers 

The exception is correct, because your models are not compatible. If you want to allow that without the exception then you must create an interface or base class from which both classes now must be derived.

The second one is not directly derive from List for a model which should be drag and dropped. This creates unexpected behavior with this library. If you still want to use it so, then you must implement your own `IDragSource` and `IDropTarget` and use a wrapper to drag and drop correctly.

```
public interface IListMember
{
    string Name { get; set; }
}

public class ListMember : IListMember
{
    public string Name { get; set; }
    public List<IListMember> SubMembers { get; set; } = new List<SubMember>();
}

public class SubMember: IListMember
{
    public string Name { get; set; }
}

public class ViewModel
{
    public ObservableCollection<IListMember> Members { get; set; } = new ObservableCollection<IListMember>();

    public ViewModel()
    {
        var listMember = new ListMember() { Name = "Item 1" };
        Members.Add(listMember);

        listMember = new ListMember() { Name = "Item 2 with SubItems" };
        listMember.SubMembers.Add(new SubMember() { Name = "Subitem1" });
        listMember.SubMembers.Add(new SubMember() { Name = "Subitem2" });
        Members.Add(listMember);
    }
}
```

Hope that helps!

**Comment by BobSammers** [comment]
Thanks for the response and for taking a look at this.

If I understand your answer, you seem to be thinking that I have tried to drag a `SubMember` into a `ListBox` of `Member`s. That is not the case (despite what the exception message implies)! I'm trying to reorder a list of `Member`s: I'm moving a `Member` into a different position in a list of `Member`s which it is already in. Apologies for not being clear enough.

My code doesn't actually show the `SubMember`s in the UI (so there is no opportunity to drag one anywhere!) - it is just the fact that one or more is present as a member of the collection exposed by a dragged `Member` that causes the behaviour I've reported.

My guess is that the code is trying to insert the `Member` and then for some reason all its `SubMember`s back into the `ListBox`'s backing collection. This presumably only occurs when the object being dropped implements IEnumerable (or one of the other collection interfaces) and definitely only occurs when that collection is populated.

I'm really regretting calling these two objects `Member` and `SubMember` now that I need to refer to them as members of a collection...
